README.md
# PRTG Zendesk Tickets Webhook

Automatically open Zendesk tickets for triggered sensors using Powershell.  Forked from [Integrating Zendesk and PRTG](https://tietze.io/b/2015/09/08/integrating-zendesk-and-prtg/)

![Zendesk Tickets](./img/tickets.png)


Before:
> Error by lookup value

After:
> Critical (2) in channel Imaging Unit (Magenta)
>
> .. if additional triggers from the same sensor, a newline to easily list the problems

Main Features:
- Automatically open Zendesk tickets when a sensor is triggered; useful for IT Staff to automate delegation to tech assistants
- Update/reopen existing tickets for the same sensor

This version has additional features:
- Tokenized authentication, to keep your password secure
- Secure API connections over TLS 1.2
- Fixes the single quote bug when using Powershell
- Option to update existing open/new tickets
- Local logging option for debugging when testing sensor messages
- A bit more newb friendly
- Formatting cleanup

![PRTG Parameters](./img/execute-program.png)

## How to use it
- First, clone/copy *ZendeskWebhook.ps1* to: `C:\Program Files (x86)\PRTG Network Monitor\Notifications\EXE` (your path may vary if using a different architecture with your PRTG server)
- Open *ZendeskWebhook.ps1* and add your config (Zendesk credentials, log options, etc; use your *API token* as the password)
- Login to your PRTG dashboard
- Setup > Account Settings > Notification Templates
- Add/click on the template you want to use with this script
- Populate the settings as you wish, then scroll toward the bottom for the **Execute Program** toggle; click it
- In the **Program File** field, select `Zendeskwebhook.ps1`
- In the **Parameters** field, add the following:
```powershell
-Device '%device' -Status '%status' -Down '%down' -Group '%group' -commentssensor '%commentssensor' -CommentsProbe '%commentsprobe' -Message '%message'
```
- In the search field (upper right corner), search for the *sensor Group* you'd like to use this script with
- Under the group list, click it > Notification Triggers > Add/edit the trigger you wish to use with this script and select the notification template you assigned the script to earlier
- Save and trigger a sensor that's using the template to test!

## Customizing
If you'd like to add additional verbiage to your tickets, you can call [additional parameters](https://www.paessler.com/manuals/prtg/list_of_placeholders_for_notifications) in similar fashion to the existing parameters.

(Don't forget to add them to *params()* to initialize inside `ZendeskWebhook.ps1`)

***
**(optional) Update Existing Tickets**

- 1) Set the `$updateExisting` variable to `1`
- 2) You can customize "update" messages by modifying the following:
  ```powershell
    # Update existing ticket or create new
    if ($existingNewTickets -gt 0 -and $updateExisting -eq 1) {

      # there is at least one open ticket for this device tagged with PRTG
      $Ticket = $SearchResults.results.Item(0)
      Write-Host "Found a ticket! Updating ticket #$($Ticket.id)"

      $Transaction = @{
        ticket = @{
          comment = @{
            public    = $false;
            body      = $CommentBody;
            author_id = $AuthorId;
          }
        }
      }
      ...
  ```
    - `public = $false;` = Whether or not you want the reply to be a public reply, or just for agents
    - `body = $CommentBody;` = If you want a custom reply, set your own message (or variable) here, otherwise, it'll re-post the initial message
    - `author_id = $AuthorId;` = If you want to change the follow-up reply author, modify that here, otherwise, the initial poster will be the reply author

### Password Auth over Token Authorization
If you prefer to use password authentication with Zendesk instead of token auth (the default), simply remove '/token' from the username.

### Troubleshooting
- Auto-closure of tickets after a device returns to Up/OK is not a feature, at this time.  If this is something you need, you'll have to extend the codebase.
- If you aren't getting tickets opened for messages that have single quotes: `Error by lookup value 'Critical (1)' in channel 'Toner (Yellow)'` - modify your contact template to have double-quotes over the **-Message** parameter, like so:
    ```text
    -Message "%lastmessage"
    ```
    *instead of*
    ```text
    -Message '%lastmessage'
    ```

### License
Tietze's release is unlicensed/public domain; my changes are licensed under GPL2


CHANGELOG.md
# PRTG Zendesk Tickets Changelog
v1.0.4
- Bugfix for multiple tickets (searches existing tickets w/ new/open status by subject, now)
- Add an option to update existing tickets

v1.0.1
- Filter 'Error by lookup value' from ticket body (leaving the important bits only!)

v1.0
- Initial release


ZendeskWebhook.ps1
# original script by https://tietze.io/b/2015/09/08/integrating-zendesk-and-prtg/
#
# solves bug: https://kb.paessler.com/en/topic/75261-single-quote-in-sensor-message-breaks-notification-script
#
# angela 11/24/21:
# - bugfix for multiple existing tickets; search by subject
# - added option to update existing ticket, or leave it be
#
# angela 12/7/18: added:
# - detailed messages & tags that don't break with single quotes
# - local log generator for easy troubleshooting
# - made the script a bit more newb friendly
# - formatting cleanup
#
# This will create a new ticket in PRTG if no open PRTG ticket is found for this device; otherwise it will update the first found open PRTG ticket for this device.

# initialize parameters that will be used later on in the script, more available at: https://kb.paessler.com/en/topic/373-what-placeholders-can-i-use-with-prtg
Param(
  [string]$Device,
  [string]$Status,
  [string]$Down,
  [string]$Group,
  [string]$CommentsSensor,
  [string]$Message,
  [string]$CommentsProbe
)

## CONFIG
# zendesk credentials
$User      = "youremail@example.com/token"  # user must be verified to use the zendesk api
$Pass      = "abcdefghijklmnopqrstuvwxyz0123456789"
# Author ID (numeric) for updates to existing tickets, ie: https://[example].zendesk.com/agent/#/users/370381808374
$AuthorId  = "1234567989"
$prtgName  = "PRTG"
$prtgEmail = "prtg@example.com"
$BaseUri   = "https://[yourzendeskurl].zendesk.com"
$updateExisting = 0
$debug     = 0  # set to 1 if you want to save a log file to the location referenced in $logPath
$logPath   = "C:\Users\Administrator\Desktop\log.txt"
## END CONFIG

### NOTE TO SELF! ###
# some of these messages, after find/replace have EXTRA spaces that get stripped from zendesk!
# copying the message from zendesk into a replace string will not work; you need to activate debug
# mode to see the true construct of the message that needs to be filtered


# remove the numeric string prefix, referenced on line 34
$Message     = $Message.Trim() -replace "^[0-9]*:",""
$Message     = $Message -replace "Error by lookup value ",""


# inject a newline when there's a dash in the comments, to keep it clean
$Message     = $Message -replace " — ","`n"


# if you want to log something other than the $Message var, simply replace the references in the conditional code block
if ($debug -eq 1 -AND (Test-Path $logPath)) {

  Write-Output $Message $CommentsSensor | Out-File $logPath -Append

} elseif ($debug -eq 1 -AND !(Test-Path $logPath)) {

  New-Item $logPath -ItemType file
  # redundant, but add-content was goobering up with unnecessary spaces between letters
  Write-Output $Message $CommentsSensor | Out-File $logPath -Append
}

$AuthHeader = @{

  "Content-Type"  = "application/json";
  "Authorization" = 'Basic ' + [Convert]::ToBase64String([Text.Encoding]::ASCII.GetBytes("$($User):$($Pass)"))
}

# call the zendesk api
function Call-ZenDesk($Api, $Method, $Body) {

  # by default, powershell seems to want to use insecure/deprecated tls, the following fixes that
  [Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12
  Invoke-RestMethod -Method $Method -Uri "$($BaseUri)$($Api)" -Headers $AuthHeader -Body $Body
}

# queue monitoring scripts don't use plural subjects, so adjust if the group match is printers, to avoid duplicate tickets
if ($($Group) -eq 'Printers') {
    $Group = 'Printer'
}

$CommentBody   = "$($CommentSsensor) $($Message) $($CommentsProbe)"
$TicketSubject = "$($Device) $($Group) Issue"
# formatting used to search tags & add new; for updating existing tickets
$deviceTag     = $Device.Trim() -replace ' ','-'

# find existing tickets, if any
$Transaction = @{

  query = "subject:$($TicketSubject) status:new status:open";
}

$SearchResults = Call-ZenDesk '/api/v2/search.json' Get $Transaction
$existingNewTickets = $($SearchResults.count)
Write-Host "Search results count: $existingNewTickets"

if ($debug -eq 1) {

  Write-Output "$existingNewTickets new/open existing tickets for: $($TicketSubject)" | Out-File $logPath -Append

}

# Update existing ticket or create new
if ($existingNewTickets -gt 0 -and $updateExisting -eq 1) {

  # there is at least one open ticket for this device tagged with PRTG
  $Ticket = $SearchResults.results.Item(0)
  Write-Host "Found a ticket! Updating ticket #$($Ticket.id)"

  $Transaction = @{
    ticket = @{
      comment = @{
        public    = $false;
        body      = $CommentBody;
        author_id = $AuthorId;
      }
    }
  }

  $Body = ConvertTo-Json($Transaction)
  Call-ZenDesk "/api/v2/tickets/$($Ticket.id).json" Put $Body

  if ($debug -eq 1) {

    Write-Output "Updated existing ticket: $($Ticket.id)" | Out-File $logPath -Append

  }

} elseif ($existingNewTickets -gt 0 -and $updateExisting -eq 0 -and $debug -eq 1) {

$existingMessage = "Existing ticket found; not updating or creating a new one."

  if ($debug -eq 1) {
    Write-Output $existingMessage | Out-File $logPath -Append
  }

} elseif ($existingNewTickets -eq 0) {

  # no ticket found, create one; to add additional tags, separate by a comma
  $Tags        = "monitoring-alert,$($deviceTag)"
  $CommentBody = "$($CommentBody)"
  $Transaction = @{

    ticket = @{
    requester = @{
      name  = "$prtgName";
      email = "$prtgEmail"
    };
    subject  = "$TicketSubject";
    type     = "Incident";
    priority = "Normal";

    comment = @{
      public = $false;
      body   = "$CommentBody";
    };
    tags = "$Tags"
    }
  }

  $Body = ConvertTo-Json $Transaction

  Call-ZenDesk "/api/v2/tickets.json" Post $Body
}
